src/mapping/query/function/split.rs

```diff
@@ -0,0 +1,183 @@
+use super::prelude::*;
+
+#[derive(Debug)]
+pub(in crate::mapping) struct SplitFn {
+    path: Box<dyn Function>,
+    pattern: ArgumentKind,
+    limit: Option<Box<dyn Function>>,
+}
+
+impl SplitFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(
+        path: Box<dyn Function>,
+        pattern: ArgumentKind,
+        limit: Option<Box<dyn Function>>,
+    ) -> Self {
+        Self {
+            path,
+            pattern,
+            limit,
+        }
+    }
+}
+
+impl Function for SplitFn {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let string = {
+            let bytes = required!(ctx, self.path, Value::Bytes(v) => v);
+            String::from_utf8_lossy(&bytes).into_owned()
+        };
+
+        let limit = optional!(ctx, self.limit, Value::Integer(i) => i)
+            .map(|limit| {
+                if limit < 0 {
+                    Err("limit is not a positive number".to_string())
+                } else {
+                    Ok(limit as usize)
+                }
+            })
+            .transpose()?;
+
+        match &self.pattern {
+            ArgumentKind::Value(path) => {
+                let pattern = required!(ctx, path, Value::Bytes(v) => v);
+                let pattern = String::from_utf8_lossy(&pattern).into_owned();
+
+                let iter: Box<dyn Iterator<Item = _>> = match limit {
+                    Some(limit) => Box::new(string.splitn(limit, &pattern)),
+                    None => Box::new(string.split(&pattern)),
+                };
+
+                Ok(Value::Array(
+                    iter.map(|sub| Value::Bytes(sub.to_string().into()))
+                        .collect(),
+                ))
+            }
+            ArgumentKind::Regex(regex) => {
+                // The global flag has no meaning here.
+                // Should we error or just ignore?
+                let iter: Box<dyn Iterator<Item = _>> = match &limit {
+                    Some(ref limit) => Box::new(regex.regex.splitn(&string, *limit)),
+                    None => Box::new(regex.regex.split(&string)),
+                };
+
+                Ok(Value::Array(
+                    iter.map(|sub| Value::Bytes(sub.to_string().into()))
+                        .collect(),
+                ))
+            }
+        }
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "pattern",
+                accepts: |v| matches!(v, Value::Bytes(_)),
```

Comment:
Am I correct in assuming that the `RemapRegex` argument kind isn't part of the validation check right now? I could see that be problematic, as that means you can now basically pass in a regex to any parameter of any function?

I assume that's not the case, so I'll go dig for where this is prevented, but it does seem like we should potentially integrate your changes with the validation logic, just so that we can keep this centralized in the exposed parameter signatures.

Reply:
Right, so I believe `ArgumentKind::into_value` now returns an error if you pass in a regex to a parameter that expects a value, correct?

I haven't thought through how to do it, but I do think it makes sense to consider updating the `accepts` check here to have it support checking for all `ArgumentKind` variants. We can optionally add some kind of macro to make the check easier to describe, e.g.:

```rust
Parameter {
	accepts: |v| accepts!(v, Regex, Value::Integer(_))
}
```

Reply:
Yes, that is correct. It was something I wasn't totally sure about, your suggestion feels more consistent so all parameter checks happen in one place.